### PR TITLE
New lint: option_needless_deref

### LINKED ISSUE #7571: Add lint checking for no-op uses of `Option::{as_deref,as_deref_mut}`
### What it does

Checks for no-op uses of `Option::{as_deref,as_deref_mut}`.

### Categories (optional)

- Kind: `complexity`

*What is the advantage of the recommended code over the original code*

Removes useless code, removing clutter and improving readability.

### Drawbacks

None.

### Example

```rust
let a = Some(&1);
let b = a.as_deref(); // goes from Option<&i32> to Option<&i32>
```

Could be written as:

```rust
let a = Some(&1);
let b = a;
```

### PR DESCRIPTION
**Pull request by lengyijun**
changelog: [`option_needless_deref`]
fix #7571

### CODE REVIEW

**Comment by camsteffen** [comment]
I think the lint name should include `as_deref`. `needless_as_deref` would be good IMO.

**Comment by lengyijun** [comment]
`option_needless_as_deref` or `needless_as_deref` ?

**Comment by camsteffen** [comment]
`needless_option_as_deref` if anything. But I'd pick `needless_as_deref`.

**Comment by lengyijun** [comment]
I did some fix.

`Result` also has a `as_deref` method, but this lint doesn't work on it. 
So I temporally rename to `needless_option_as_deref`.

File: clippy_lints/src/needless_option_as_deref.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+]);
+
+impl<'tcx> LateLintPass<'tcx> for OptionNeedlessDeref {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+        if in_macro(expr.span) {
+            return;
+        }
```
Comment (llogiq) [inline clippy_lints/src/needless_option_as_deref.rs:46]:
```suggestion
        if expr.span.from_expansion() || in_macro(expr.span) {
            return;
        }
```

**Review by llogiq — APPROVED** [review]
I only have a small nit, otherwise this looks good.

**Comment by llogiq** [comment]
Thank you! @bors r+

**Pull request merged**